### PR TITLE
#3046 persistent failed item status and #2065 failed items reporting error once

### LINKED ISSUE #2065: nuclei should not fail silently
I ran into an issue where my template was clearly not executed for no apparent reasons (https://github.com/projectdiscovery/nuclei/discussions/2062)

It turns out the template was not run after reaching too many errors.

I found this out using the `-stats` flag and my workaround was to use the `-max-host-error` flag with a high value.

I believe when the max-host-error limit is reached (default 30) there should be a clear error message.

### LINKED ISSUE #3046: hosterrorscache erroneously reset max failed counter
### Nuclei version:
dev

### Current Behavior:
In `hosterrorscache.go`, each item in the cache is set to -1 after reaching the maximum threshold. This makes it ineffective, as the host is again processed later. Adding a print instruction at https://github.com/projectdiscovery/nuclei/blob/b355b76afdcc3a2216ff71988636bc206713935c/v2/pkg/protocols/common/hosterrorscache/hosterrorscache.go#L99:
image

### Expected Behavior:
The counter shouldn't be reset and the host invalidated permanently. Probably it's enough to remove the counter reset. Additional requirements:
- [ ] The output info line, when a host gets inserted in the error cache, should be printed only once
- [ ] Implement tests to ensure that the cache behaves properly

### Example
add the following snippet in `hosterrorcache.go` (function `Check`):
```go
if numberOfErrors == -1 {
	return true
}
log.Println(finalValue, numberOfErrorsValue)
if numberOfErrorsValue >= c.MaxHostError {
	_ = c.failedTargets.Set(finalValue, -1)
	if c.verbose {
		gologger.Verbose().Msgf("Skipping %s as previously unresponsive %d times", finalValue, numberOfErrorsValue)
	}
	return true
}
log.Println(finalValue, numberOfErrorsValue)
```

Execute the following command:
```console
$ echo http://asdasjkdashkjdahsjkdhas | go run . -elog testlog.txt -duc -no-interactsh
```

### PR DESCRIPTION
**Pull request by xm1k3**
## Proposed changes
This PR:
- Fixes #3046 by removing the set to -1 behaviour
- Fixes #2065 by implementing sync.Once behavior on self-evicting cache items 

## Checklist
- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [x] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
- [ ] Failing tests
- [ ] Implement test cases, minimum cases to cover:
    - [ ] host inserted into the cache and multiple async calls to MarkFailed don't trigger counter reset
    - [ ] multiple calls to `cacheItem.Do()` are executed only once

File: v2/pkg/protocols/common/hosterrorscache/hosterrorscache_test.go [hunk 1]
```diff
@@ -10,18 +10,69 @@ import (
...
+
+		value, ok := failedTarget.(*cacheItem)
+		require.True(t, ok)
+		require.EqualValues(t, test.expected, value.errors.Load())
 	}
-	cache.MarkFailed("example.com", fmt.Errorf("could not resolve host"))
-	if value, err := cache.failedTargets.Get("example.com"); err == nil && value != nil {
-		require.Equal(t, 1, value, "could not get correct number of marked failed hosts")
+
+	for i := 0; i < 3; i++ {
```
Comment (Mzack9999) [inline v2/pkg/protocols/common/hosterrorscache/hosterrorscache_test.go:34]: This should be moved into an independent test, the test goal is to ensure that after the max error threshold the item is still there with the expected amount of errors correctly counted

File: v2/pkg/protocols/common/hosterrorscache/hosterrorscache_test.go [hunk 2]
```diff
@@ -10,18 +10,69 @@ import (
...
+		cache.MarkFailed("test", fmt.Errorf("could not resolve host"))
 	}
+
+	value := cache.Check("test")
+	require.Equal(t, true, value, "could not get checked value")
+}
+
+func TestCacheItemCheckMarkFailedMultipleCalls(t *testing.T) {
```
Comment (Mzack9999) [inline v2/pkg/protocols/common/hosterrorscache/hosterrorscache_test.go:42]: The purpose of this test should be to call the `cache.MarkFailed` multiple times in parallell from various go routines on the same item, and like my previous comment, we need to ensure that the counter remains above the max error threshold.

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

File: v2/pkg/protocols/common/hosterrorscache/hosterrorscache_test.go [hunk 3]
```diff
@@ -2,30 +2,112 @@ package hosterrorscache
...
+		existingCacheItem, err := cache.failedTargets.GetIFPresent(normalizedCacheValue)
+		require.Nil(t, err)
+		require.NotNil(t, existingCacheItem)
+		existingCacheItemValue := existingCacheItem.(*cacheItem)
+		skippingValue := existingCacheItemValue.errors.Load() >= int32(test.expected)
+		require.Equal(t, true, skippingValue, "Didn't skipped host")
+	}
+}
+
+func TestCacheMarkFailedConcurrent(t *testing.T) {
```
Comment (Mzack9999) [inline v2/pkg/protocols/common/hosterrorscache/hosterrorscache_test.go:79]: The test should look like the following one:
```go
func TestCacheMarkFailedConcurrent(t *testing.T) {
	cache := New(3, DefaultMaxHostsCount)

	tests := []struct {
		host     string
		expected int32
	}{
		{"http://example.com:80", 200},
		{"example.com:80", 200},
		{"example.com", 100},
	}

	wg := sync.WaitGroup{}
	for _, test := range tests {
		currentTest := test
		for i := 0; i < 100; i++ {
			wg.Add(1)

			go func() {
				defer wg.Done()

				cache.MarkFailed(currentTest.host, fmt.Errorf("could not resolve host"))
			}()
		}
	}

	wg.Wait()

	for _, test := range tests {
		require.True(t, cache.Check(test.host))

		normalizedCacheValue := cache.normalizeCacheValue(test.host)
		failedTarget, err := cache.failedTargets.Get(normalizedCacheValue)
		require.Nil(t, err)
		require.NotNil(t, failedTarget)

		value, ok := failedTarget.(*cacheItem)
		require.True(t, ok)
		require.EqualValues(t, test.expected, value.errors.Load())
	}
}
```
The purpose is to ensure that concurrent calls to `cache.MarkFailed` generate the final expected result (`cache.Check` true and corrent amount of errors. double for equal normalized values)

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
We also need the additional test with multiple concurrent calls to `cacheItem..Do(..)` ensuring that the function is executed only once. Something similar to this in pseudocode:
```go
var (
	count int
	item     cacheItem
)

wg := sync.waitgroup{}
for i := 0; i < xxx; i++ {
        wg.Add(1)
        go func() {
		defer wg.Done()
		item.Do(func() {
			count++
		})
	} ()
}
wg.wait()

require.Equal(t, count, 1) // ensures the increment happened only once regardless of the multiple call
```

File: v2/pkg/protocols/common/hosterrorscache/hosterrorscache_test.go [hunk 4]
```diff
@@ -2,30 +2,112 @@ package hosterrorscache
...
+		require.Nil(t, err)
+		require.NotNil(t, failedTarget)
+
+		value, ok := failedTarget.(*cacheItem)
+		require.True(t, ok)
+		require.EqualValues(t, test.expected, value.errors.Load())
+	}
+}
+
+func TestCacheMarkFailedMultipleCalls(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (xm1k3) [reply v2/pkg/protocols/common/hosterrorscache/hosterrorscache_test.go:47]: I changed the name of the test and the values of the hosts

**Review by Mzack9999 — APPROVED** [review]
lgtm - Asking @ehsandeep review due to go1.19 version bump

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**